hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/RowStatisticsCompactionObserver.java

```diff
@@ -0,0 +1,321 @@
+        NAMESPACED_TABLE_NAME, e);
+    }
+  }
+
+  @Override
+  public InternalScanner preCompact(ObserverContext<RegionCoprocessorEnvironment> context,
+    Store store, InternalScanner scanner, ScanType scanType, CompactionLifeCycleTracker tracker,
+    CompactionRequest request) {
+    if (RowStatisticsUtil.isInternalTable(store.getTableName())) {
+      LOG.debug("Region {} belongs to an internal table {}, so no row statistics will be recorded",
+        store.getRegionInfo().getRegionNameAsString(), store.getTableName().getNameAsString());
+      return scanner;
+    }
+    int blocksize = store.getColumnFamilyDescriptor().getBlocksize();
+    boolean isMajor = request.isMajor();
+    RowStatisticsImpl stats;
+    if (isMajor) {
+      stats = new RowStatisticsImpl(store.getTableName().getNameAsString(),
+        store.getRegionInfo().getEncodedName(), store.getColumnFamilyName(), blocksize,
+        maxCacheSize, true);
+    } else {
+      stats = new RowStatisticsImpl(store.getTableName().getNameAsString(),
+        store.getRegionInfo().getEncodedName(), store.getColumnFamilyName(), blocksize,
+        maxCacheSize, false);
+    }
+    return new RowStatisticsScanner(scanner, isMajor, stats, context.getEnvironment());
+  }
+
+  private class RowStatisticsScanner implements InternalScanner, Shipper {
+
+    private final InternalScanner scanner;
+    private final Shipper shipper;
+    private final boolean isMajor;
+    private final RowStatisticsImpl rowStatistics;
+    private final RegionCoprocessorEnvironment regionEnv;
+    private RawCellBuilder cellBuilder;
+    private Cell lastCell;
+
+    public RowStatisticsScanner(InternalScanner scanner, boolean isMajor,
+      RowStatisticsImpl rowStatistics, RegionCoprocessorEnvironment regionEnv) {
+      this.scanner = scanner;
+      if (scanner instanceof Shipper) {
+        this.shipper = (Shipper) scanner;
+      } else {
+        this.shipper = null;
+      }
+      this.isMajor = isMajor;
+      this.rowStatistics = rowStatistics;
+      this.regionEnv = regionEnv;
+      this.cellBuilder = regionEnv.getCellBuilder();
+    }
+
+    @Override
+    public boolean next(List<? super ExtendedCell> list, ScannerContext scannerContext)
+      throws IOException {
+      boolean ret = scanner.next(list, scannerContext);
+
+      if (list.isEmpty()) {
+        return ret;
+      }
+
+      // each next() call returns at most 1 row (maybe less for large rows)
+      // so we just need to check if the first cell has changed rows
+      ExtendedCell first = (ExtendedCell) list.get(0);
+      if (rowChanged(first)) {
+        rowStatistics.handleRowChanged(lastCell);
+      }
+
+      for (int i = 0; i < list.size(); i++) {
+        ExtendedCell cell = (ExtendedCell) list.get(i);
+        rowStatistics.consumeCell(cell);
+        lastCell = cell;
+      }
+
+      return ret;
+    }
+
+    @Override
+    public boolean next(List<? super ExtendedCell> result) throws IOException {
```

Comment:
Why does this version of the method not perform any accounting?

Reply:
That was an oversight on my part. The interface changed from branch-2.5 (what we're running at my day job) and master (here)